### PR TITLE
Tracing messages pretty morpho

### PR DESCRIPTION
**Pull request by koslambrou**
Made tracing message more human readable. Here are examples:

```
[papineau:morpho.chain-db:Debug:7] [2021-04-16 14:32:32.30 UTC] Immutable DB: No valid last location

[papineau:morpho.extract-state:Info:45] [2021-04-16 14:32:32.30 UTC] Current Ledger State: MorphoBlock (lastCheckpoint: Checkpoint(block: PowBlock(number: 0, hash: )), checkpointAt: Genesis (Origin), currentVotes: (), morphoTip: 7927d2fa@13429523)

[papineau:morpho.consensus.forge:Info:34] [2021-04-16 14:32:32.30 UTC] Forged block for slot 13429523: MorphoBlock (hash: 7927d2fa, prevHash: GenesisHash, slot: 13429523, blockNo: 0, body: MorphoBody())

[papineau:morpho.consensus.forge:Info:34] [2021-04-16 14:32:38.00 UTC] Forged block for slot 13429529: MorphoBlock (hash: 28acd92f, prevHash: BlockHash(974abd9c), slot: 13429529, blockNo: 6, body: MorphoBody(MorphoBlockTx(transaction: Vote(PowBlock(number: 4, hash: 5250128d)), transactionId: 8cd0c0ca)))

[papineau:morpho.consensus.mempool:Info:47] [2021-04-16 14:32:37.31 UTC] Transaction Vote(PowBlock(number: 4, hash: 5250128d)) added to mempool. Size of mempool goes from 0 to 1. Bytes of mempool goes from 0 to 2000

[papineau:morpho.consensus.mempool:Debug:32] [2021-04-16 14:32:38.01 UTC] Transactions were removed because they're no longer valid: [ Vote(PowBlock(number: 4, hash: 5250128d)) ]. Mempool size is 0 transactions and 0 bytes

[papineau:morpho.chain-db:Debug:34] [2021-04-16 14:32:32.30 UTC] Block added to queue: (Point 13429523, "7927d2fab8883300"). Queue size 1

[papineau:morpho.chain-db:Debug:24] [2021-04-16 14:32:32.30 UTC] Chain added block (Point 13429523, "7927d2fab8883300")

[papineau:morpho.consensus.chain-sync-server-header:Debug:72] [2021-04-16 14:32:32.32 UTC] Synchronizing chain: current tip is Tip(slot: 13429523, hash: 7927d2fa, blockNo: 0). Chain update by rolling back to 7927d2fa@13429523

[papineau:morpho.consensus.chain-sync-server-header:Debug:72] [2021-04-16 14:32:33.00 UTC] Synchronizing chain: roll forward to 87171813@13429524

[papineau:morpho.consensus.chain-sync-client:Debug:89] [2021-04-16 14:32:32.32 UTC] Found the intersection point 7927d2fa@13429523 between our chain fragment Tip(slot: 13429523, hash: 7927d2fa, blockNo: 0) and the candidate's chain Tip(slot: 13429523, hash: 7927d2fa, blockNo: 0)

[papineau:morpho.ip-subs:Debug:57] [2021-04-16 14:32:32.30 UTC] IPs: 0.0.0.0:0 [::]:0 [127.0.0.1:3000,127.0.0.1:3002,127.0.0.1:3004] Waiting 0.025s before attempting a new connection

[papineau:morpho.extract-state:Debug:45] [2021-04-16 14:32:35.00 UTC] Checkpoint won't be pushed: Checkpoint block Genesis (Origin) is not the current Morpho tip 7927d2fa@13429523

[papineau:morpho.rpc:Info:47] [2021-04-16 14:32:37.31 UTC] GetLatestBlock(request: (CheckpointingInterval: 4, PowBlock(number: 0, hash: )), event: RpcStart)

[papineau:morpho.rpc:Info:45] [2021-04-16 14:32:41.00 UTC] PushCheckpoint(request: Checkpoint(block: PowBlock(number: 4, hash: 5250128d)), event: RpcStart)

[papineau:morpho.consensus.block-fetch-decision:Debug:36] [2021-04-16 14:32:34.00 UTC] Peer ConnectionId {localAddress = 127.0.0.1:46139, remoteAddress = 127.0.0.1:3000}: Blocks not fetched from peer: chain not plausible

[papineau:morpho.tx-submission-protocol-2:Debug:81] [2021-04-16 14:32:32.31 UTC] Peer ConnectionId {localAddress = 127.0.0.1:46139, remoteAddress = 127.0.0.1:3000}: Send message: (ClientAgency TokHello,MsgHello)

[papineau:morpho.tx-submission-protocol-2:Debug:80] [2021-04-16 14:32:32.31 UTC] Peer ConnectionId {localAddress = 0.0.0.0:3000, remoteAddress = 127.0.0.1:46139}: Receive message: (ClientAgency TokHello,MsgHello)

[papineau:morpho.error-policy:Notice:55] [2021-04-16 14:32:32.33 UTC] 127.0.0.1:3002: Suspending consumer with given exception: Just (ConnectionExceptionTrace Network.Socket.connect: <socket: 24>: does not exist (Connection refused))

[papineau:morpho.ip-subs:Debug:64] [2021-04-16 14:32:32.30 UTC] IPs: 0.0.0.0:0 [::]:0 [127.0.0.1:3000,127.0.0.1:3002,127.0.0.1:3004] Connection Attempt Start, destination 127.0.0.1:3000

[papineau:morpho.ip-subs:Debug:57] [2021-04-16 14:32:32.30 UTC] IPs: 0.0.0.0:0 [::]:0 [127.0.0.1:3000,127.0.0.1:3002,127.0.0.1:3004] Trying to connect to 127.0.0.1:3000

[papineau:morpho.ip-subs:Debug:64] [2021-04-16 14:32:32.30 UTC] IPs: 0.0.0.0:0 [::]:0 [127.0.0.1:3000,127.0.0.1:3002,127.0.0.1:3004] Connection Attempt End, destination 127.0.0.1:3000, outcome: ConnectSuccess

[papineau:morpho.mux:Debug:67] [2021-04-16 14:32:32.30 UTC] Mux bearer ConnectionId {localAddress = 0.0.0.0:3000, remoteAddress = 127.0.0.1:46139} with event: Bearer Receive Header Start

```

### CODE REVIEW

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 1]
```diff
@@ -132,8 +262,25 @@ instance
   where
   mpretty = viaShow
 
-instance (StandardHash blk) => MPretty (TraceChainSyncServerEvent blk) where
-  mpretty = viaShow
+instance MPretty (GenTx (MorphoBlock h c)) where
+  mpretty (MorphoGenTx tx _) = mpretty tx
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:266]: The second parameter should also be printed, in order for being able to easily correlate `TraceTxSubmissionOutbound` traces

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 2]
```diff
@@ -349,37 +707,143 @@ instance (Show tx, Show txid) => MPretty (TraceTxSubmissionOutbound txid tx) whe
...
+  mpretty (WithIPList (NtN.LocalAddresses Nothing (Just ipv6) Nothing) wilDsts wilEvent) =
+    "IPs:" <+> viaShow ipv6 <+> viaShow wilDsts <+> mpretty wilEvent
+  mpretty (WithIPList (NtN.LocalAddresses (Just ipv4) Nothing Nothing) wilDsts wilEvent) =
+    "IPs:" <+> viaShow ipv4 <+> viaShow wilDsts <+> mpretty wilEvent
+  mpretty (WithIPList (NtN.LocalAddresses Nothing Nothing (Just unix)) wilDsts wilEvent) =
+    "IPs:" <+> viaShow unix <+> viaShow wilDsts <+> mpretty wilEvent
+  mpretty (WithIPList (NtN.LocalAddresses (Just ipv4) (Just ipv6) Nothing) wilDsts wilEvent) =
+    "IPs:" <+> viaShow ipv4 <+> viaShow ipv6 <+> viaShow wilDsts <+> mpretty wilEvent
+  mpretty WithIPList {wilSrc, wilDsts, wilEvent} =
+    "IPs:" <+> viaShow wilSrc <+> viaShow wilDsts <+> mpretty wilEvent
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:817]: This instance can be improved. I'm thinking

> src: [\<ip1\>, \<ip2\>], dst: [\<ip1\>, \<ip2\>], event: \<event\>

You can turn a `LocalAddresses addr` into `[addr]` with
```haskell
(LocalAddresses a b c) = catMaybes [a, b, c]
```

then you only need a single pattern match

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 3]
```diff
@@ -349,37 +707,143 @@ instance (Show tx, Show txid) => MPretty (TraceTxSubmissionOutbound txid tx) whe
...
   mpretty xs = list $ map mpretty xs
 
-instance (StandardHash blk) => MPretty (ChainHash blk) where
-  mpretty = viaShow
+instance (MPretty a) => MPretty (Maybe a) where
+  mpretty = mpretty
+
+instance (MPretty a, MPretty b) => MPretty (Either a b) where
+  mpretty (Left e) = mpretty e
+  mpretty (Right v) = mpretty v
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:830]: Whether it's left or right is important, so you should annotate it e.g. with `"Left(" <> mpretty e <> ")"`, same for `Right`

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 4]
```diff
@@ -349,37 +707,143 @@ instance (Show tx, Show txid) => MPretty (TraceTxSubmissionOutbound txid tx) whe
...
   mpretty = viaShow
 
-instance Condense (HeaderHash blk) => MPretty (Point blk) where
-  mpretty pt =
-    case pt of
-      GenesisPoint -> "genesis (origin)"
-      BlockPoint slot h -> pretty $ condense h ++ "@" ++ condense slot
+instance MPretty (HeaderHash blk) => MPretty (Point blk) where
+  mpretty GenesisPoint = "Genesis (Origin)"
+  mpretty (BlockPoint slot h) = mpretty h <> "@" <> mpretty slot
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:846]: I think we should make this more clearer, I've always been confused by the `@` in the logs. I'm thinking `"Block(slot:" <+> mpretty slot <> ", hash:" <+> h <> ")"`

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 5]
```diff
@@ -349,37 +707,143 @@ instance (Show tx, Show txid) => MPretty (TraceTxSubmissionOutbound txid tx) whe
...
   where
   mpretty = viaShow
 
-instance Condense (HeaderHash blk) => MPretty (Point blk) where
-  mpretty pt =
-    case pt of
-      GenesisPoint -> "genesis (origin)"
-      BlockPoint slot h -> pretty $ condense h ++ "@" ++ condense slot
+instance MPretty (HeaderHash blk) => MPretty (Point blk) where
+  mpretty GenesisPoint = "Genesis (Origin)"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:845]: We can shorten this to just `Genesis`, it's clear enough like that.

File: morpho-checkpoint-node/src/Morpho/RPC/Abstract.hs [hunk 6]
```diff
@@ -66,13 +77,44 @@ instance ToJSON ev => ToJSON (RpcTrace ev i o) where
...
+      <+> mpretty request
+      <> ", event:"
+      <+> viaShow event
+      <> ")"
+  mpretty (RpcTrace PushCheckpoint request event) =
+    "PushCheckpoint(request:"
+      <+> mpretty request
+      <> ", event:"
+      <+> viaShow event
+      <> ")"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/RPC/Abstract.hs:91]: This can have a nicer instance, currently it looks like

> GetLatestBlock(request: (CheckpointingInterval: 4, PowBlock(number: 236, hash: b45e6ad6)), event: RpcSuccess Nothing)

I think this would be much nicer e.g. like

> RPC call GetLatestBlock(interval: 4, parent: PowBlock(number: 236, hash: b45e6ad6)) returned: Nothing

For pushing:

> RPC call PushCheckpoint(Checkpoint(block: PowBlock(number: 236, hash: b45e6ad6))) returned: true

And for the `RpcStart` event:

> Starting RPC call GetLatestBlock(interval: 4, parent: PowBlock(number: 236, hash: b45e6ad6))


In general I think that all the `Trace` instances (like `SubscriptionTrace`, `RpcTrace`, etc.) should have a sentence-like structure. On the other hand, data structures like `MorphoBlock`, `Vote`, etc. should have a structural instance like `MorphoBlock(hash: ...)`. These structural instances can be used in the `Trace` instances.

Comment (Infinisil) [reply morpho-checkpoint-node/src/Morpho/RPC/Abstract.hs:91]: You can do a nested pattern match on the `event` to implement this

**Review by Infinisil — CHANGES_REQUESTED** [review]

File: morpho-checkpoint-node/src/Morpho/RPC/Abstract.hs [hunk 7]
```diff
@@ -46,8 +45,20 @@ data RpcTrace ev i o
   = RpcTrace (RpcMethod i o) i (RpcEvent ev o)
 
 instance Show ev => Show (RpcTrace ev i o) where
-  show (RpcTrace method@GetLatestBlock request event) = "RpcTrace " <> Prelude.show method <> " " <> Prelude.show request <> " " <> Prelude.show event
-  show (RpcTrace method@PushCheckpoint request event) = "RpcTrace " <> Prelude.show method <> " " <> Prelude.show request <> " " <> Prelude.show event
+  show (RpcTrace method@GetLatestBlock request event) =
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/RPC/Abstract.hs:48]: I don't think we should need any `Show` instances anymore, so this should be removed

File: morpho-checkpoint-node/src/Morpho/RPC/Abstract.hs (hunk 6 above)
Comment (koslambrou) [reply morpho-checkpoint-node/src/Morpho/RPC/Abstract.hs:91]: @Infinisil If I understand correctly, I should only use the `RpcSuccess` data constructor for the event and leave out `RpcStart` and `RpcEvent` ?

Comment (Infinisil) [reply morpho-checkpoint-node/src/Morpho/RPC/Abstract.hs:91]: No, to have all of those be encoded as sentences, rather than structures. So for example:
```haskell
mpretty (RpcTrace GetLatestBlock (k, parent) RpcStart) = "Starting RPC call GetLatestBlock(interval:" <+> mpretty k <> ", parent:" <+> mpretty parent
mpretty (RpcTrace PushCheckpoint block (RpcSuccess result)) = "Success, RPC call PushCheckpoint(" <> mpretty block <> " returned" <+> mpretty result
```

And that for all combinations of the method and event

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 8]
```diff
@@ -88,13 +138,66 @@ instance
...
+      <> ", blockNo:"
+      <+> mpretty b
+      <> ")"
 
 instance MPretty Tx where
   mpretty (Tx vote) = mpretty vote
 
 instance MPretty Vote where
-  mpretty = viaShow
+  mpretty (Vote block _) = mpretty block
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:167]: We can also use `recoverPublicKey` here to know which node signed it. I'd go for a format like `Vote(block: \<block\>, byPubkey: \<pubkey\>)`

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 9]
```diff
@@ -315,71 +478,463 @@ instance
...
+  where
+  mpretty (WithMuxBearer peer a) = "Mux bearer" <+> mpretty peer <+> "with:" <+> mpretty a
+
 instance MPretty a => MPretty [a] where
-  mpretty xs = list $ map mpretty xs
+  -- mpretty xs = list $ map mpretty xs
+  mpretty xs =
+    "["
+      <> concatWith (\d1 d2 -> d1 <> "," <+> d2) (map mpretty xs)
+      <> "]"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:904]: Check out how prettyprinter's [`prettyList`](https://hackage.haskell.org/package/prettyprinter-1.7.0/docs/Prettyprinter.html#v:prettyList) is implemented. We can do the same here as a simplification

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 10]
```diff
@@ -88,13 +139,73 @@ instance
...
+
+instance MPretty Checkpoint where
+  mpretty (Checkpoint powBlockRef signatures) =
+    let bytes = powBlockRefToBytes powBlockRef
+        pubKeys = map (`recoverPublicKey` bytes) signatures
+     in "Checkpoint(block:"
+          <+> mpretty powBlockRef
+          <> ", signedByPubkeys:"
+          <+> mpretty pubKeys
+          <> ")"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:195]: Currently this is of type `[Maybe PublicKey]`, which then prints as

> [Just(PublicKey("3ea6ebf3")), Just(PublicKey("ec33a368"))]

We can make this a bit nicer with

```haskell
instance MPretty Checkpoint where
  mpretty (Checkpoint powBlockRef signatures) =
    let bytes = powBlockRefToBytes powBlockRef
        showSignature sig = maybe "<unknown>" mpretty (recoverPublicKey sig bytes)
        pubKeys = map showSignature signatures
     in "Checkpoint(block:"
          <+> mpretty powBlockRef
          <> ", signedByPubkeys:"
          <+> list pubKeys
          <> ")"
```

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 11]
```diff
@@ -1,81 +1,132 @@
...
-      MorphoStdHeader {..} = morphoHeaderStd
-      MorphoBody {..} = morphoBody
 
-instance
-  (HashAlgorithm h, BftCrypto c) =>
-  MPretty (Header (MorphoBlock h c))
-  where
+instance MPretty PublicKey where
+  mpretty (PublicKey bytes) =
+    "PublicKey(" <> viaShow (take 8 $ Prelude.show bytes) <> ")"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:103]: This currently prints `"` around the bytes.

I suggest:
- Creating a `Morpho.Common.Bytes.bytesToHex :: Bytes -> Text`, which is based on the `Show Bytes` implementation
- Using `bytesToHex` for `Show Bytes`
- Using `bytesToHex` to implement `MPretty Bytes`
- Then using `mpretty bytes` in the implementation of `MPretty PublicKey`

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 12]
```diff
@@ -315,71 +478,463 @@ instance
...
+  mpretty (ChainRange p1 p2) = "(" <> mpretty p1 <> "," <+> mpretty p2 <> ")"
 
-instance (Show tx, Show txid) => MPretty (TraceTxSubmissionOutbound txid tx) where
-  mpretty = viaShow
+instance
+  ( forall (st :: ps). Show (ClientHasAgency st),
+    forall (st :: ps). Show (ServerHasAgency st),
+    forall (st :: ps) (st' :: ps). Show (Message ps st st')
+  ) =>
+  MPretty (NtN.TraceSendRecv ps)
```
[the remark this answers is not in the document]
Comment (Infinisil) [reply morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:756]: A lot of these instances are similar to the ones in `OrphanToJSONInstances.hs`. E.g. I'm thinking this one will be necessary too:
```haskell
instance BftCrypto c => MPretty (SerialisedHeader (MorphoBlock h c)) where
  mpretty (SerialisedHeaderFromDepPair (GenDepPair (NestedCtxt CtxtMorpho) b)) = mpretty b
```

**Review by Infinisil — CHANGES_REQUESTED** [review]

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs (hunk 9 above)
Comment (koslambrou) [reply morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:904]: @Infinisil The problem with `prettyList`'s implementation is that it uses `list`. That function will automatically create line breaks for every element of the list when the line has more than `n` characters (default is 80). Example of an output:

```
[1
, 2
, 3]
```

Of course, one solution would be to set the number of characters to a much higher value. 

What do you think?

Comment (koslambrou) [reply morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:904]: @Infinisil  `prettyList` uses the `list` function which might introduce line breaks. However, I just found the `hsep` function which doesn't

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs (hunk 11 above)
Comment (koslambrou) [reply morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:103]: Just found a `bytesToHex` in Morpho.Common.Conversions`. However, I can't use that function in `Show Bytes` since it introduces a cyclic dependency. But I can use it for tracing.

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 13]
```diff
@@ -1,81 +1,159 @@
...
 {-# LANGUAGE RecordWildCards #-}
 {-# LANGUAGE UndecidableInstances #-}
 
 module Morpho.Tracing.Pretty where
 
 import Cardano.Crypto.DSIGN.Class (Signable)
-import Cardano.Crypto.Hash.Class (Hash, HashAlgorithm, hashToTextAsHex)
+import Cardano.Crypto.Hash.Class (Hash, hashToTextAsHex)
 import Cardano.Prelude hiding (list)
+import Cardano.Slotting.Slot (WithOrigin (At, Origin))
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:18]: The `cardano-slotting` library has been removed as a dependency, but we can get the same imports from

```haskell
import Ouroboros.Consensus.Block.Abstract (WithOrigin(..))
```

instead

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 14]
```diff
@@ -1,81 +1,159 @@
...
+instance MPretty MorphoBody where
+  mpretty (MorphoBody txs) =
+    "MorphoBody("
+      <> hsep (punctuate comma (map mpretty txs))
+      <> ")"
+
+instance MPretty (Header (MorphoBlock h c)) where
   mpretty MorphoHeader {..} =
     "("
       <> mpretty morphoPrev
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:159]: This instance should be improved, an example of how it looks is

> (BlockHash(3e867db0)->c337db48,13948056)

which is very ambiguous. Let's use the same format as the others instead.

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 15]
```diff
@@ -315,71 +491,768 @@ instance
...
+        <+> pretty numTxsAfter
+        <> ". Bytes of mempool goes from"
+        <+> pretty numBytesBefore
+        <+> "to"
+        <+> pretty numBytesAfter
+  mpretty (TraceMempoolRejectedTx tx err (MempoolSize numTxs numBytes)) =
+    "Transaction"
+      <+> mpretty tx
+      <+> "rejected because:"
+      <+> mpretty err <> ". Mempool has"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:707]:
```suggestion
      <+> mpretty (map snd err) <> ". Mempool has"
```

Type of `err` is `(Vote, MorphoTransactionError)`, so it carries the vote with it, but in this message we already have the vote in the `tx`, so we can only look at the second value here.

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 16]
```diff
@@ -1,81 +1,159 @@
...
+instance MPretty (HeaderHash blk) => MPretty (MorphoState blk) where
+  mpretty (MorphoState lastCkpt ckptAt votes tip) =
+    "MorphoState(morphoTip:"
+      <+> mpretty tip
+      <> ", lastCheckpoint:"
+      <+> mpretty lastCkpt
+      <> ", checkpointAt:"
+      <+> mpretty ckptAt
+      <> ", currentVotes:"
+      <+> mpretty votes
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:122]: I'm now seeing that some information is duplicated here:

> currentVotes: (PublicKey(ec33a368) -> Vote(block: PowBlock(number: 4, hash: 1b3eb035), byPubkey: Just(PublicKey(ec33a368))))

So I think let's go for just printing the list of signatures here, so

```suggestion
      <+> mpretty (Map.elems votes)
```

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 17]
```diff
@@ -88,13 +166,74 @@ instance
...
 
 instance MPretty Vote where
-  mpretty = viaShow
+  mpretty (Vote block signature) =
+    let bytes = powBlockRefToBytes block
+        pubKey = recoverPublicKey signature bytes
+     in "Vote(block:"
+          <+> mpretty block
+          <> ", byPubkey:"
+          <+> mpretty pubKey
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:201]: Currently this is

> Vote(block: PowBlock(number: 4, hash: 1b3eb035), byPubkey: Just(PublicKey(ec33a368))

Let's get rid of the `Just` the same way as for `MPretty Checkpoint`, showing `Nothing` as `\<unknown\>`.

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 18]
```diff
@@ -88,13 +166,74 @@ instance
...
+  mpretty (PowBlockHash bytes) = mpretty bytes
+
+instance MPretty PowBlockRef where
+  mpretty (PowBlockRef number hash) =
+    "PowBlock(number:" <+> mpretty number <> ", hash:" <+> mpretty hash <> ")"
+
+instance MPretty Checkpoint where
+  mpretty (Checkpoint powBlockRef signatures) =
+    let bytes = powBlockRefToBytes powBlockRef
+        showSignature sig = maybe "<unknown>" mpretty (recoverPublicKey sig bytes)
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:217]: We can reuse this for `MPretty Vote` by making this a global function

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 19]
```diff
@@ -315,71 +489,773 @@ instance
...
-      GenesisPoint -> "genesis (origin)"
-      BlockPoint slot h -> pretty $ condense h ++ "@" ++ condense slot
+instance MPretty (HeaderHash blk) => MPretty (Point blk) where
+  mpretty GenesisPoint = "Genesis"
+  mpretty (BlockPoint slot h) =
+    "Block(slot:" <+> mpretty slot <> ", hash:" <+> mpretty h <> ")"
+
+instance MPretty (HeaderHash blk) => MPretty (RealPoint blk) where
+  mpretty (RealPoint slot headerHash) =
+    "RealPoint(slot:" <+> mpretty slot <> ", hash:" <+> mpretty headerHash <> ")"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:1251]: Let's just use `Block(slot:` here, the fact that this is a `RealPoint` isn't useful.

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 20]
```diff
@@ -1,100 +1,237 @@
...
-      MorphoStdHeader {..} = morphoHeaderStd
-      MorphoBody {..} = morphoBody
 
-instance
-  (HashAlgorithm h, BftCrypto c) =>
-  MPretty (Header (MorphoBlock h c))
-  where
+instance MPretty PublicKey where
+  mpretty (PublicKey bytes) =
+    "PublicKey(" <> mpretty bytes <> ")"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:127]: And here let's change this to just

```suggestion
    mpretty bytes
```

The fact that it's a `PublicKey` is always obvious from context.

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 21]
```diff
@@ -1,100 +1,237 @@
...
+  mpretty (MorphoBlockTx genTx genTxId) =
+    "MorphoBlockTx(transaction:"
+      <+> mpretty genTx
+      <> ", transactionId:"
+      <+> mpretty genTxId
+      <> ")"
+
+instance MPretty (ChainHash (MorphoBlock h c)) where
+  mpretty GenesisHash = "GenesisHash"
+  mpretty (BlockHash h) = "BlockHash(" <> mpretty h <> ")"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:178]:
```suggestion
  mpretty GenesisHash = "Genesis"
  mpretty (BlockHash h) = mpretty h
```

More dense information this way

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 22]
```diff
@@ -105,35 +242,75 @@ instance MPretty SlotNo where
...
 
-instance (StandardHash blk) => MPretty (TraceChainSyncServerEvent blk) where
-  mpretty = viaShow
+instance MPretty (GenTx (MorphoBlock h c)) where
+  mpretty (MorphoGenTx tx morphoTxId) =
+    "Transaction(tx:"
+      <+> mpretty tx
+      <> ", morphoTxId:"
+      <+> mpretty morphoTxId
+      <> ")"
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:296]:
```suggestion
  mpretty (MorphoGenTx tx morphoTxId) =
    "Transaction(id:"
      <+> mpretty morphoTxId
      <> ", tx:"
      <+> mpretty tx
      <> ")"
```

Id makes more sense before the contents.

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 23]
```diff
@@ -1,100 +1,237 @@
...
-instance
-  (HashAlgorithm h, BftCrypto c) =>
-  MPretty (Header (MorphoBlock h c))
-  where
+instance MPretty PublicKey where
+  mpretty (PublicKey bytes) =
+    "PublicKey(" <> mpretty bytes <> ")"
+
+instance MPretty Bytes where
+  mpretty = pretty . Text.take 8 . bytesToHex
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:130]: I think we should add `0x` prefixes to hexified things.

```suggestion
  mpretty = pretty . ("0x" <>) . Text.take 8 . bytesToHex
```

File: morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs [hunk 24]
```diff
@@ -1,100 +1,237 @@
...
+      <+> mpretty p1
+      <+> "is not the current Morpho tip"
+      <+> mpretty p2
+
+instance MPretty VoteError where
+  mpretty (FailedToSignBlockRef powBlockRef) =
+    "Failed to sign block reference" <+> mpretty powBlockRef
 
 instance MPretty (Hash h a) where
   mpretty = pretty . Text.take 8 . hashToTextAsHex
```
Comment (Infinisil) [inline morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs:237]:
```suggestion
  mpretty = pretty . ("0x" <>) . Text.take 8 . hashToTextAsHex
```

**Review by Infinisil — COMMENTED** [review]
I hope you don't mind the nitpicks, but I think after these small changes it's looking good!

**Review by Infinisil — APPROVED** [review]
